### PR TITLE
Add basic functional S3 tests for standalone NSFS using boto3 

### PR DESCRIPTION
**Pull request by sagihirshfeld**
Besides the tests that were added in `test_s3_bucket_operations.py` and `test_s3_object_operations.py`, a few additional important changes were made:

1. The `account_manager` `s3_client_factory` usage in `conftest.py` has been changed in a way that we can now use the in the class and session scopes as well. This allows saving time on redundant setups for test suites that don't make any changes to NSFS accounts or service settings, like is the case in `test_s3_bucket_operations.py` and `test_s3_object_operations.py`

2. `NSFSAccount.create()` changes:

- More arguments have been added to the method, allowing more configurable account creation. The need for this came up while testing S3 bucket creation using an account with the `allow_buckets_creation` property set to false. The required adjustments to existing usage of the method have been made.

- Default value generation options have been added for the `account_name`, `secret_key` and `prefix_key` arguments, and the `create` method now returns a tuple with these values. Since all of the existing usage so far we had to repeatedly call the `generate_unique_resource_name` and `generate_random_hex` util functions to generate the values for these arguments from outside the class, usually with the default arguments of these util functions, moving their usage to the `create` method itself saves the need for repeated calls while allowing the more customizable approach when needed.

### CODE REVIEW

File: tests/test_s3_bucket_operations.py [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+        1. Create a bucket via S3
+        2. Verify the bucket was created via S3 HeadBucket
+        3. Delete the bucket via S3
+        4. Verify the bucket was deleted via S3 HeadBucket
+
+        """
+        # 1. Create a bucket via S3
+        bucket_name = c_scope_s3client.create_bucket()
+        # 2. Verify the bucket was created via S3 HeadBucket
+        assert c_scope_s3client.head_bucket(bucket_name), "Bucket was not created"
```
Comment (udaysk23) [inline tests/test_s3_bucket_operations.py:29]: Please check the response code to validate bucket creation

File: tests/test_s3_bucket_operations.py [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+        # 3. Delete the bucket via S3
+        c_scope_s3client.delete_bucket(bucket_name)
+        # 4. Verify the bucket was deleted via S3 HeadBucket
+        assert (
+            c_scope_s3client.head_bucket(bucket_name) == False
+        ), "Bucket was not deleted"
```
Comment (udaysk23) [inline tests/test_s3_bucket_operations.py:35]: Please check the response code to validate bucket deletion

File: noobaa_sa/s3_client.py [hunk 3]
```diff
@@ -102,15 +130,47 @@ def delete_bucket(self, bucket_name, empty_before_deletion=False):
...
+            log.info(f"head_bucket call to {bucket_name} returned 200 (OK)")
+            return True
+        elif response_code == 404:
+            log.warn(f"head_bucket call to {bucket_name} returned 404 (Not Found)")
+            return False
+        else:
+            raise UnexpectedBehaviour(
+                f"Unexpected response from head_bucket call to {bucket_name}: {response}"
+            )
+
```
Comment (udaysk23) [inline noobaa_sa/s3_client.py:173]: To standardize s3 operations in this file, Please return the response directory and user will validate the return codes based on their test need

File: noobaa_sa/s3_client.py [hunk 4]
```diff
@@ -145,21 +209,37 @@ def list_objects(self, bucket_name, prefix="", use_v2=False):
...
+        try:
+            self._boto3_client.put_object(Bucket=bucket_name, Key=object_key, Body=body)
+        except ClientError as e:
+            if e.response["Error"]["Code"] == "NoSuchBucket":
+                log.warn(
+                    f"Bucket {bucket_name} does not exist and cannot be written to"
+                )
+                raise NoSuchBucket(e)
+            else:
+                raise e
```
Comment (udaysk23) [inline noobaa_sa/s3_client.py:242]: There might be situation where we need to verify the operations based on response code. I would recommend here to return response dictionary for these boto3 methods

File: noobaa_sa/s3_client.py [hunk 5]
```diff
@@ -177,9 +257,21 @@ def get_object(self, bucket_name, object_key):
...
-        output = self._boto3_client.get_object(Bucket=bucket_name, Key=object_key)
+        try:
+            output = self._boto3_client.get_object(Bucket=bucket_name, Key=object_key)
+        except ClientError as e:
+            if e.response["Error"]["Code"] == "NoSuchKey":
+                log.warn(f"Object {object_key} does not exist in bucket {bucket_name}")
+                raise NoSuchKey(e)
+            else:
+                raise e
+
```
Comment (udaysk23) [inline noobaa_sa/s3_client.py:274]: Same as mentioned above

File: noobaa_sa/s3_client.py [hunk 6]
```diff
@@ -241,10 +334,82 @@ def delete_object(self, bucket_name, object_key):
...
+                    f"Source or destination bucket does not exist: {e.response['Error']['Message']}"
+                )
+                raise NoSuchBucket(e)
+            elif e.response["Error"]["Code"] == "NoSuchKey":
+                log.warn(
+                    f"Source object {src_key} does not exist in bucket {src_bucket}"
+                )
+                raise NoSuchKey(e)
+            else:
+                raise e
```
Comment (udaysk23) [inline noobaa_sa/s3_client.py:411]: Same for this as well

File: tests/test_s3_bucket_operations.py [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+
+log = logging.getLogger(__name__)
+
+
+class TestS3BucketOperations:
+    """
+    Test S3 bucket operations using NSFS
+    """
+
+    def test_bucket_creation_deletion_and_head(self, c_scope_s3client):
```
Comment (udaysk23) [inline tests/test_s3_bucket_operations.py:17]: Can we have head_bucket as separate operation in same test

File: tests/test_s3_bucket_operations.py [hunk 8]
```diff
@@ -0,0 +1,189 @@
...
+        try:
+            for _ in range(AMOUNT):
+                buckets.append(c_scope_s3client.create_bucket())
+
+            listed_buckets = c_scope_s3client.list_buckets()
+
+            # listed_buckets might contain buckets from before the test
+            assert all(
+                bucket in listed_buckets for bucket in buckets
+            ), "Created bucket was not listed!"
```
Comment (udaysk23) [inline tests/test_s3_bucket_operations.py:53]: Please use set operation here

File: tests/test_s3_bucket_operations.py (hunk 1 above)
Comment (sagihirshfeld) [reply tests/test_s3_bucket_operations.py]: Done in https://github.com/red-hat-storage/noobaa-sa-ci/pull/16/commits/027ec7aac1aaa31c2625b0649ff6bf65372da000

File: tests/test_s3_bucket_operations.py (hunk 2 above)
Comment (sagihirshfeld) [reply tests/test_s3_bucket_operations.py]: Done in https://github.com/red-hat-storage/noobaa-sa-ci/commit/3c9c0dce3ae21b8c2d2006d988a66e1c8d539af9

File: noobaa_sa/s3_client.py (hunk 3 above)
Comment (sagihirshfeld) [reply noobaa_sa/s3_client.py]: Done in https://github.com/red-hat-storage/noobaa-sa-ci/commit/3c9c0dce3ae21b8c2d2006d988a66e1c8d539af9

Note that boto3 returns a ClientError with a different response dict in the error, with a different path to the response code. In my implementation I'm returning either this or the "successful" dict, and to allow uniform access of the response code, I also add it under the "Code" property in the root of the dict.

File: noobaa_sa/s3_client.py [hunk 9]
```diff
@@ -68,11 +83,19 @@ def create_bucket(self, bucket_name=""):
...
         """
         if bucket_name == "":
             bucket_name = generate_unique_resource_name(prefix="bucket")
         log.info(f"Creating bucket {bucket_name} via boto3")
-        response = self._boto3_client.create_bucket(Bucket=bucket_name)
+        try:
+            response = self._boto3_client.create_bucket(Bucket=bucket_name)
+        except ClientError as e:
+            if e.response["Error"]["Code"] == "BucketAlreadyExists":
+                raise BucketAlreadyExists(e)
```
[the remark this answers is not in the document]
Comment (sagihirshfeld) [reply noobaa_sa/s3_client.py]: Done in https://github.com/red-hat-storage/noobaa-sa-ci/pull/16/commits/027ec7aac1aaa31c2625b0649ff6bf65372da000

File: noobaa_sa/s3_client.py (hunk 4 above)
Comment (sagihirshfeld) [reply noobaa_sa/s3_client.py]: Done in https://github.com/red-hat-storage/noobaa-sa-ci/pull/16/commits/cde85fc0c8aae17993097f4a72b721a00cc42873

File: noobaa_sa/s3_client.py (hunk 5 above)
Comment (sagihirshfeld) [reply noobaa_sa/s3_client.py]: Done in https://github.com/red-hat-storage/noobaa-sa-ci/pull/16/commits/cde85fc0c8aae17993097f4a72b721a00cc42873

File: noobaa_sa/s3_client.py (hunk 6 above)
Comment (sagihirshfeld) [reply noobaa_sa/s3_client.py]: Done in https://github.com/red-hat-storage/noobaa-sa-ci/pull/16/commits/cde85fc0c8aae17993097f4a72b721a00cc42873

File: tests/test_s3_bucket_operations.py (hunk 8 above)
Comment (sagihirshfeld) [reply tests/test_s3_bucket_operations.py]: Done in https://github.com/red-hat-storage/noobaa-sa-ci/pull/16/commits/7f271a08a8fe91de47b2d8e10e4d553a0ed051d4

**Comment by sagihirshfeld** [comment]
Passing logs link after the latest changes: https://url.corp.redhat.com/a479cd6

File: tests/test_s3_bucket_operations.py (hunk 7 above)
Comment (sagihirshfeld) [reply tests/test_s3_bucket_operations.py:20]: Per our discussion, the new implementation of this makes use of the return codes of head_bucket as well so it's sufficient - resolving this conversation

File: tests/test_s3_bucket_operations.py [hunk 10]
```diff
@@ -0,0 +1,205 @@
...
+                "Listed object last modified time is not within a reasonable range",
+                f"Object: {written}, Last Modified: {last_modified}",
+            )
+            # 4.c. Verify that the sizes match
+            expected_size = os.path.getsize(os.path.join(origin_dir, written))
+            listed_size = listed["Size"]
+            assert expected_size == listed_size, (
+                "Listed object size does not match written object size",
+                f"Object: {written}, Expected: {expected_size}, Actual: {listed_size}",
+            )
```
Comment (udaysk23) [inline tests/test_s3_bucket_operations.py:150]: Please move this test method in tests/test_s3_object_operations.py file

**Review by udaysk23 — CHANGES_REQUESTED** [review]

File: noobaa_sa/account.py [hunk 11]
```diff
@@ -62,18 +67,47 @@ class NSFSAccount(Account):
...
+
+        # Set default values if not provided
+        if not account_name:
+            account_name = generate_unique_resource_name(prefix="account")
+        if not access_key:
+            access_key = generate_random_hex()
+        if not secret_key:
+            secret_key = generate_random_hex()
+        if not email:
+            account_email = config.ENV_DATA["email"]
```
Comment (udaysk23) [inline noobaa_sa/account.py:110]: As per the latest build, Access_key should contains Uppercase and lower case characters.
Please refer https://github.com/red-hat-storage/noobaa-sa-ci/pull/19

Comment (sagihirshfeld) [reply noobaa_sa/account.py]: resolved in https://github.com/red-hat-storage/noobaa-sa-ci/pull/16/commits/de02bcff175fb2a11114113c0c5786da75a7f636

File: tests/test_s3_bucket_operations.py (hunk 10 above)
Comment (sagihirshfeld) [reply tests/test_s3_bucket_operations.py]: According to the docs the listing of objects is considered as an operation we make on the bucket: https://www.ibm.com/docs/en/ssdas?topic=overview-s3-buckets

**Comment by sagihirshfeld** [comment]
Passing logs of all the tests after the latest changes: https://privatebin.corp.redhat.com/?1f528c8ab3c5be6a#958hYXGRgTRt732wSJJjmspKP1ZLsrRtCWZL1ooQajx6

File: tests/test_s3_bucket_operations.py [hunk 12]
```diff
@@ -0,0 +1,207 @@
...
+
```
Comment (udaysk23) [inline tests/test_s3_bucket_operations.py:153]: As requested earlier, Please move this test to tests/test_s3_object_operations.py

**Comment by sagihirshfeld** [comment]
Passing logs after the latest change to `test_put_and_get_obj`: https://privatebin.corp.redhat.com/?cb300b24d69f3434#HgTMmryVa9EXioY2zz5Lf2fzNZnPwX4a15x24y1ZADBJ

**Review by udaysk23 — APPROVED** [review]

**Review by vavuthu — APPROVED** [review]

**Pull request merged**